File: C++/21. 재귀/4. 24060.cpp
#include <iostream>

using namespace std;

int count = 0, k;
int tmp[500000];
bool print = false;

void merge(int *a, int p, int q, int r) {
    int i = p, j = q + 1, t = 1;

    while (i <= q && j <= r) {
        if (a[i] <= a[j])
            tmp[t++] = a[i++];
        else
            tmp[t++] = a[j++];
    }

    while (i <= q)
        tmp[t++] = a[i++];
        
    while (j <= r)
        tmp[t++] = a[j++];

    i = p;
    t = 1;
    while (i <= r) {
        a[i++] = tmp[t++]; 
        count++;
        if (count == k) {
            cout << a[i - 1];
            print = true;
        }
    }
}

void merge_sort(int *a, int p, int r) {
    if (p < r) {
        int q = (p + r) / 2;
        merge_sort(a, p, q);
        merge_sort(a, q + 1, r);
        merge(a, p, q, r);
    }
}

int main() {
    ios::sync_with_stdio(false);
    cin.tie(NULL);
    cout.tie(NULL);

    int n, a[500000];
    cin >> n >> k;

    for (int i = 0; i < n; i++)
        cin >> a[i];

    merge_sort(a, 0, n - 1);

    if (!print)
        cout << -1;

    return 0;
}

File: 백준 2회차/21. 재귀/4. 24060.py
import sys

def merge_sort(A: list, p: int, r: int):
    if p < r:
        q = (p + r) // 2
        merge_sort(A, p, q)
        merge_sort(A, q + 1, r)
        merge(A, p , q, r)

def merge(A: list, p: int, q: int, r: int):
    i, j, t = p, q + 1, 1
    tmp = []

    while i <= q and j <= r:
        if A[i] <= A[j]:
            tmp.append(A[i])
            i += 1
        else:
            tmp.append(A[j])
            j += 1

    while i <= q:
        tmp.append(A[i])
        i += 1
    while j <= r:
        tmp.append(A[j])
        j += 1
    
    i, t = p, 0
    global count, result
    while i <= r:
        count += 1
        if count == k:
            result = tmp[t]
        A[i] = tmp[t]
        i += 1
        t += 1

n, k = map(int, sys.stdin.readline().rstrip().split())
A = list(map(int, sys.stdin.readline().rstrip().split()))
count = 0
result = 0
merge_sort(A, 0, n - 1)
if count >= k:
    print(result)
else:
    print(-1)